command/operator_migrate.go

```diff
@@ -0,0 +1,303 @@
+	"github.com/hashicorp/hcl/hcl/ast"
+	"github.com/hashicorp/vault/command/server"
+	"github.com/hashicorp/vault/helper/logging"
+	"github.com/hashicorp/vault/physical"
+	"github.com/mitchellh/cli"
+	"github.com/pkg/errors"
+	"github.com/posener/complete"
+)
+
+var _ cli.Command = (*OperatorMigrateCommand)(nil)
+var _ cli.CommandAutocomplete = (*OperatorMigrateCommand)(nil)
+
+type OperatorMigrateCommand struct {
+	*BaseCommand
+
+	PhysicalBackends map[string]physical.Factory
+	flagConfig       string
+	flagStart        string
+	flagReset        bool
+	logger           log.Logger
+}
+
+type migratorConfig struct {
+	StorageSource      *server.Storage `hcl:"-"`
+	StorageDestination *server.Storage `hcl:"-"`
+}
+
+func (c *OperatorMigrateCommand) Synopsis() string {
+	return "Migrates Vault data between storage backends"
+}
+
+func (c *OperatorMigrateCommand) Help() string {
+	helpText := `
+Usage: vault operator migrate [options]
+
+  This command starts a storage backend migration process to copy all data
+  from one backend to another. This operates directly on encrypted data and
+  does not require a Vault server, nor any unsealing.
+
+  Start a migration with a configuration file:
+
+      $ vault operator migrate -config=migrate.hcl
+
+  For more information, please see the documentation.
+
+` + c.Flags().Help()
+
+	return strings.TrimSpace(helpText)
+}
+
+func (c *OperatorMigrateCommand) Flags() *FlagSets {
+	set := NewFlagSets(c.UI)
+	f := set.NewFlagSet("Command Options")
+
+	f.StringVar(&StringVar{
+		Name:   "config",
+		Target: &c.flagConfig,
+		Completion: complete.PredictOr(
+			complete.PredictFiles("*.hcl"),
+		),
+		Usage: "Path to a configuration file. This configuration file should " +
+			"contain only migrator directives.",
+	})
+
+	f.StringVar(&StringVar{
+		Name:   "start",
+		Target: &c.flagStart,
+		Usage:  "Only copy keys lexicographically at or after this value.",
+	})
+
+	f.BoolVar(&BoolVar{
+		Name:   "reset",
+		Target: &c.flagReset,
+		Usage:  "Reset the migration lock. No migration will occur.",
+	})
+
+	return set
+}
+
+func (c *OperatorMigrateCommand) AutocompleteArgs() complete.Predictor {
+	return nil
+}
+
+func (c *OperatorMigrateCommand) AutocompleteFlags() complete.Flags {
+	return c.Flags().Completions()
+}
+
+func (c *OperatorMigrateCommand) Run(args []string) int {
+	c.logger = logging.NewVaultLogger(log.Info)
+	f := c.Flags()
+
+	if err := f.Parse(args); err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	if c.flagConfig == "" {
+		c.UI.Error("Must specify exactly one config path using -config")
+		return 1
+	}
+
+	config, err := c.loadMigratorConfig(c.flagConfig)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error loading configuration from %s: %s", c.flagConfig, err))
+		return 1
+	}
+
+	if err := c.migrate(config); err != nil {
+		c.UI.Error(fmt.Sprintf("Error migrating: %s", err))
+		return 2
+	}
+
+	if c.flagReset {
+		c.UI.Output("Success! Migration lock reset (if it was set).")
```

Comment:
Now that we are not using the lock to gate startup of the server i wonder if it's needed at all? Since it's not a background process i wonder if we can just let it run to completion without locking, and it enables two migration to two different storage backends at once. What do you think?

Reply:
We are still using the lock to gate server startup (https://github.com/hashicorp/vault/pull/5143/commits/d0daeee8d381bf345b5ca12cd52e5ec336367205#diff-593cd0d199cb70fe727e6b64eb150baaR470). The most recent change in that area was to remove the ability to override that lock. The `-reset` option is mainly for if the process died/was killed and a stale lock exists.

We _could_ look at loosening the restriction on starting a second migration (to a different destination) though. I can put that down as an future enhancement, along with the idea of parallelized migration.